src/gui/properties/piecesmapwidget.cpp

```diff
@@ -0,0 +1,195 @@
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "piecesmapwidget.h"
+
+#include <QtMath>
+#include <QPainter>
+#include <QResizeEvent>
+#include <QScrollBar>
+#include <QProgressBar>
+#include <QProxyStyle>
+#include <QStyleOptionProgressBar>
+
+#include "base/bittorrent/torrent.h"
+#include "propertieswidget.h"
+#include "base/bittorrent/session.h"
+#include "base/bittorrent/infohash.h"
+
+
+PiecesMapWidget::PiecesMapWidget(QScrollArea *parent)
+    : QWidget(parent)
+    , m_scrollArea(parent)
+{
+    m_piecesCount = 0;
+    m_piecePixelSize = 15;
+    m_spacePixelSize = 5;
+    m_height = 0;
+
+    BitTorrent::Session* session = BitTorrent::Session::instance();
+    QObject::connect(session, &BitTorrent::Session::pieceFinished, this, &PiecesMapWidget::handlePieceFinished);
+
+    QProgressBar dummyProgressBar;
+    dummyProgressBar.setFixedSize(m_piecePixelSize, m_piecePixelSize);
+    dummyProgressBar.setTextVisible(false);
+    dummyProgressBar.setMinimum(0);
+    dummyProgressBar.setMaximum(100);
+#if (defined(Q_OS_WIN) || defined(Q_OS_MACOS))
+    auto *fusionStyle = new QProxyStyle {"fusion"};
+    fusionStyle->setParent(&dummyProgressBar);
+    dummyProgressBar.setStyle(fusionStyle);
+#endif
+    QPixmap pixmap(m_piecePixelSize, m_piecePixelSize);
+    QPalette palette = dummyProgressBar.palette();
+    palette.setColor(QPalette::Highlight, Qt::transparent);
+    dummyProgressBar.setValue(100);
+    dummyProgressBar.render(&pixmap);
+    m_pieceFinishedImage = pixmap.toImage();
+    dummyProgressBar.setPalette(palette);
+    dummyProgressBar.render(&pixmap);
+    m_pieceNeededImage = pixmap.toImage();
+}
+
+PiecesMapWidget::~PiecesMapWidget()
+{
+}
+
+QSize PiecesMapWidget::sizeHint() const
+{
+    QSize hint = QWidget::sizeHint();
+    if(m_torrent == nullptr) return hint;
+    hint.setHeight(m_height);
+    return hint;
+}
+
+QSize PiecesMapWidget::minimumSizeHint() const
+{
+    QSize hint = QWidget::minimumSizeHint();
+    if(m_torrent == nullptr) return hint;
+    hint.setHeight(m_height);
+    return hint;
+}
+
+int PiecesMapWidget::heightForWidth(int width) const
+{
+    int height = QWidget::heightForWidth(width);
+    if(m_torrent == nullptr) return height;
+    float viewportWidth = (float)m_scrollArea->viewport()->width();
+    float piecesPerLine = viewportWidth / (float)(m_piecePixelSize + m_spacePixelSize);
+    int lines = qCeil((float)m_piecesCount / piecesPerLine);
+    height = lines * (m_piecePixelSize + m_spacePixelSize);
+    return height;
+}
+
+void PiecesMapWidget::clear()
+{
+    m_pieces.clear();
+}
+
+void PiecesMapWidget::handlePieceFinished(BitTorrent::Torrent *torrent)
+{
+    if(!m_torrent || torrent->id().toString() != m_torrent->id().toString()) return;
+    this->update();
```

Comment:
Why not `update();`?

Reply:
If you mean why I used this->, it's just an idiomatic. Fixed

Reply:
What kind of idiom are you talking about?

Reply:
I always add "this->" in my code even when it's not necessary for some reasons (mainly to quickly focus on what is a member class and what is not). I used the word idiomatic but it's not what I meant, the correct word should be idiot-(auto)matic, in french that sound the same.